src/enzyme_ad/jax/Passes/AffineToStableHLORaising.cpp

```diff
@@ -944,6 +978,217 @@ emitStoreAsScatter(Location loc, Value update, Value input, ValueRange sIndices,
+  }
+
+  for (auto memref : entryBlock->getArguments()) {
+    Value mappedMemref = mapping.lookup(memref);
+    inits.push_back(mappedMemref);
+
+    cond->addArgument(mappedMemref.getType(), mappedMemref.getLoc());
+    Value memrefInBody =
+        body->addArgument(mappedMemref.getType(), mappedMemref.getLoc());
+    mapping.map(memref, memrefInBody);
+  }
+
+  auto whileOp = builder.create<stablehlo::WhileOp>(forOp.getLoc(), inits);
+
+  whileOp->getRegion(0).push_back(cond);
+  whileOp->getRegion(1).push_back(body);
+
+  {
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointToStart(cond);
+    Value cond = builder.create<stablehlo::CompareOp>(
+        forOp.getLoc(), ivInCond, ub, stablehlo::ComparisonDirection::LT);
+    builder.create<stablehlo::ReturnOp>(forOp.getLoc(), cond);
+  }
+  {
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointToStart(body);
+
+    mapping.map(iv, ivInBody);
+    maps[ivInBody] =
+        affine::AffineValueMap(AffineMap::get(forOp->getContext()), {});
+
+    for (auto &innerOp : forOp.getBody()->without_terminator()) {
+      if (tryRaisingOpToStableHLO(&innerOp, mapping, builder, maps, pc)
+              .failed())
+        return failure();
+    }
+
+    Value newIvInBody =
+        builder.create<stablehlo::AddOp>(forOp.getLoc(), ivInBody, step);
+
+    SmallVector<Value> loopCarried = {newIvInBody};
+    for (auto [iterArg, yieldedIterArgs] :
+         llvm::zip(forOp.getRegionIterArgs(),
+                   forOp.getBody()->getTerminator()->getOperands())) {
+      if (maps.lookup(mapping.lookup(iterArg)) !=
+          maps.lookup(mapping.lookup(yieldedIterArgs))) {
+        LLVM_DEBUG(llvm::dbgs() << "invalid init for iterArg: ";
+                   iterArg.printAsOperand(llvm::dbgs(), OpPrintingFlags());
+                   llvm::dbgs() << "\n");
+        return failure();
+      }
+      loopCarried.push_back(mapping.lookup(yieldedIterArgs));
+    }
+
+    for (auto memref : entryBlock->getArguments())
+      loopCarried.push_back(mapping.lookup(memref));
+    builder.create<stablehlo::ReturnOp>(forOp.getLoc(), loopCarried);
+  }
+
+  for (auto [i, memref] : llvm::enumerate(entryBlock->getArguments()))
+    mapping.map(memref,
+                whileOp.getResult(i + 1 + forOp.getNumRegionIterArgs()));
+  for (auto [forRes, forIterArg, whileRes] :
+       llvm::zip(forOp.getResults(), forOp.getRegionIterArgs(),
+                 llvm::drop_begin(whileOp.getResults()))) {
+    mapping.map(forRes, whileRes);
+    maps[whileRes] = maps.lookup(mapping.lookup(forIterArg));
+  }
+
+  return success();
+}
+
+template <class T> static SmallVector<BlockArgument, 6> getIVs(T op);
+template <> SmallVector<BlockArgument, 6> getIVs(affine::AffineParallelOp op) {
+  return {op.getIVs().begin(), op.getIVs().end()};
+}
+template <> SmallVector<BlockArgument, 6> getIVs(affine::AffineForOp op) {
+  return {op.getInductionVar()};
+}
```

Comment:
Isn't `LoopLikeInterface::getLoopInductionVars` working  for these?

Reply:
I found the std::optional wrapping of the result of that function confusing so I opted to use these.